lib/Sema/TypeCheckCodeCompletion.cpp

```diff
@@ -1309,6 +1281,42 @@ sawSolution(const constraints::Solution &S) {
   Results.push_back({ExpectedTy, SingleExprBody});
 }
 
+void addTypeToResultVectorIfNecessary(SmallVectorImpl<Type> &Vec, Type Ty) {
+  // Don't add null or unresolved types because they don't produce useful
+  // results.
+  if (!Ty || Ty->is<UnresolvedType>()) {
+    return;
+  }
+
+  // Don't add types twice.
+  if (llvm::any_of(Vec,
+                   [&](const Type &TyInVec) { return TyInVec->isEqual(Ty); })) {
+    return;
+  }
+  Vec.push_back(Ty);
+}
+
+void PatternMatchingTypeCheckCompletionCallback::sawSolution(
+    const constraints::Solution &S) {
+  GotCallback = true;
+
+  Type CompletionExprType = getTypeForCompletion(S, CompletionExpr);
+  addTypeToResultVectorIfNecessary(Results.ExprPatternTypes,
+                                   CompletionExprType);
+
+  if (auto MatchVar = PatternMatching->getMatchVar()) {
```

Comment:
Can't we derive `MatchVar` or `MatchingVarType` from `Solution`? If not, could you make it so?
If we can do that, I think we don't need to add a dedicated completion kind. Instead, we can detect it in `UnresolvedMemberTypeCheckCompletionCallback::sawSolution()`

Reply:
Not easily, I think. We could maybe search through the constraint locators for one that declares a variable named `$match`, but that feels awfully ad-hoc to me and potentially fragile if `$match` somehow leaks to nested scopes.

Also note that we only offer enum cases for the type of the match var, so `UnresolvedMemberTypeCheckCompletionCallback` would need to keep track of the match var types separately from the current `Results`.

We could merge `PatternMatchingTypeCheckCompletionCallback` and `UnresolvedMemberTypeCheckCompletionCallback` so that `UnresolvedMemberTypeCheckCompletionCallback` also has `ExprPattern *PatternMatching` and `SmallVector<Type, 1> EnumPatternTypes` members, which can be `nullptr`/empty if we are not completing in a pattern matching location. I’m not sure if I like that better though… WDYT?